tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,209 @@
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use std::future::Future;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use tokio::runtime::Builder;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::{spawn_local, JoinHandle, LocalSet};
+
+/// A handle to a local pool, used for spawning `!Send` tasks.
+#[derive(Clone)]
+pub struct LocalPoolHandle {
+    pool: Arc<LocalPool>,
+}
+
+impl LocalPoolHandle {
+    /// Create a new pool of threads to handle `!Send` tasks. Spawn tasks onto this
+    /// pool via [`LocalPoolHandle::spawn_pinned`].
+    ///
+    /// # Panics
+    /// Panics if the pool size is less than one.
+    pub fn new(pool_size: usize) -> LocalPoolHandle {
+        assert!(pool_size > 0);
+
+        let workers = (0..pool_size)
+            .map(|_| LocalWorkerHandle::new_worker())
+            .collect();
+
+        let pool = Arc::new(LocalPool { workers });
+
+        LocalPoolHandle { pool }
+    }
+
+    /// Spawn a task onto a worker thread and pin it there so it can't be moved
+    /// off of the thread. Note that the future is not [`Send`], but the
+    /// [`FnOnce`] which creates it is.
+    ///
+    /// This function returns a future because the `create_task` function must
+    /// be sent to a worker thread and spawned there, so we need to
+    /// (asynchronously) wait for the join handle to be sent back. The task is
+    /// sent to the pool immediately, so if you don't need the join handle you
+    /// can drop the future.
+    ///
+    /// # Examples
+    /// ```
+    /// use std::rc::Rc;
+    /// use tokio_util::task::LocalPoolHandle;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     // Create the local pool
+    ///     let pool = LocalPoolHandle::new(1);
+    ///
+    ///     // Spawn a !Send future onto the pool and await it
+    ///     let output = pool
+    ///         .spawn_pinned(|| {
+    ///             // Rc is !Send + !Sync
+    ///             let local_data = Rc::new("test");
+    ///
+    ///             // This future holds an Rc, so it is !Send
+    ///             async move { local_data.to_string() }
+    ///         })
+    ///         // First await is to get the JoinHandle
+    ///         .await // JoinHandle<String>
+    ///         // Second await is to get the task output
+    ///         .await // Result<String, JoinError>
+    ///         .unwrap();
+    ///
+    ///     assert_eq!(output, "test");
+    /// }
+    /// ```
+    pub fn spawn_pinned<F, Fut>(
+        &self,
+        create_task: F,
+    ) -> impl Future<Output = JoinHandle<Fut::Output>>
```

Comment:
I can't think of much more besides improving this return type. Some options:

 * Create a custom `JoinHandle` type for our local set.
 * Keep around a `Handle` to the runtime, and `spawn` on it remotely - this can't be used to spawn a local task, but the task could in turn spawn a local task on the `LocalSet` and forward the return value to the other `JoinHandle`. It would be a bit funky, but would work and _would_ let us return a `JoinHandle` immediately without blocking.

Reply:
I experimented with the second option, using the runtime handle. However, in my first attempt the `spawn_local` call panics because it's not called in the context of the local set (in the task spawned by `block_on`).

Also, since `LocalSet` is `!Send + !Sync` we can't pass around a handle to it to directly call its `spawn_local` method. (This was an idea that I thought of, but isn't possible.)

Here's a gist of the code I used for the first attempt:
https://gist.github.com/AzureMarker/99d7447a3b8e2e206a1ad52c0e30884d

I tried again, keeping the channel infrastructure, and I got some thing that works pretty well (commit is pushed). I also added in a good amount of error handling and comments. I realized there are still a few issues though around error handling.

One issue is that panicking in `create_task` before returning the future will kill the worker. I tried using `catch_unwind` to avoid this, but that limits the range of closures you can pass to `spawn_pinned`. For example, you can't use an `std::sync::mpsc` channel because it's not unwind safe (this shows up in the `can_drop_future_and_still_get_output` test). We could use `AssertUnwindSafe` here, but we probably shouldn't (?). I didn't include this in the commit I pushed, but here is the code:
https://gist.github.com/AzureMarker/f6ff1033eb45602b78e0c09660ac5311

The second issue is that, if the worker does die for some reason, the pool will continue to use the worker handle. This will cause random `spawn_pinned` failures down the line. Ideally, the pool should recreate the worker when it sees a failure. However, that will introduce mutable state to the pool and require some form of mutual exclusion when repairing the pool.

Reply:
There's no problem with using `AssertUnwindSafe` in that situation.

As for the Tokio runtime, well it just catches all panics so that workers don't die.

Reply:
Well I guess that's a simple solution. I pushed the change to use `catch_unwind` and `AssertUnwindSafe`.